### PR TITLE
CI: Add scripts to run samples automatically

### PR DESCRIPTION
**Pull request by AntonioND**
The scripts are a mix of bash and a bit of GNU expect for flexibility, as each sample needs to be tested in slightly different ways.

Each sample must have a test.sh script in its folder so that the CI detects it and runs it. They are executed by run_sample.sh script, created using run_test.sh as an example.

This PR adds several samples to the CI, but not all of them:

- The ml folder samples, for example, take far too long to be part of a regular CI run, and need caching of docker images, which is a task on its own.

- The openmp and nodejs samples seem to be broken.

This PR also fixes the following samples:

- ``languages/dotnet``: The following error happened when trying to run it:

```
FailFast:
Couldn't find a valid ICU package installed on the system. Set the configuration flag
System.Globalization.Invariant to true if you want to run with no globalization support.
```

By adding a new file with this configuration flag set to true, the sample works again. https://github.com/dotnet/core/issues/2186#issuecomment-472629489

- ``basic/helloworld``: This PR also simplifies the sample so that it matches the helloworld test.

- ``languages/java``:  When the number of ethreads is 1 the java app hangs after printing the message. With a higher number of ethreads it exits as expected.

This is a partial fix for https://github.com/lsds/sgx-lkl/issues/231, and is intended to make https://github.com/lsds/sgx-lkl/issues/332 easier by adding this new bash/expect framework.

### CODE REVIEW

**Comment by AntonioND** [comment]
I need to make this work in the CI, it looks like the build output isn't saved where I expected it to be.

**Comment by AntonioND** [comment]
So now the 8 samples I've added to the CI pass (the CI failures are not this PR's fault).

They also pass in my VM (with ``run-hw`` and ``run-sw``):

```
run_mode=run-sw \
SGXLKL_ETHREADS=1 \
SGXLKL_BUILD_MODE=debug \
SGXLKL_ROOT=~/sgx-lkl \
.azure-pipelines/scripts/test_runner.sh samples  
```

File: samples/basic/attack/read_memory.sh
Comment (SeanTAllen) [inline samples/basic/attack/read_memory.sh]: I believe our standard is always "; then" for bash if/then formatting. Based on what I've seen in the project.

File: samples/containers/encrypted/test.sh
Comment (SeanTAllen) [inline samples/containers/encrypted/test.sh:14]: is there an issue for this TODO or is it going to be added to an existing issue?

File: samples/containers/encrypted/test.sh
Comment (SeanTAllen) [inline samples/containers/encrypted/test.sh]: wow. out of curiosity, how long does this test usually take?

File: samples/containers/redis/README.md
Comment (SeanTAllen) [inline samples/containers/redis/README.md]: I think the list should start at 1.

This "0" is weird to me.

File: samples/languages/dotnet/README.md
Comment (SeanTAllen) [inline samples/languages/dotnet/README.md:15]: Do we normally give two options with no explanation of which you should do? If yes, then I think this is ok, because it isn't worse than what we usually do, however, unless someone knows what run-hw and run-sw do, this provides no context for picking one.

I think it should have a short explanation for each "run with SGX on hardware that supports it" or "run emulated". Something to that effect.

File: samples/languages/java/README.md
Comment (SeanTAllen) [inline samples/languages/java/README.md]: same for "0" rather than "1"

File: samples/languages/java/test.sh
Comment (SeanTAllen) [inline samples/languages/java/test.sh]: looks like you left a blank line here accidentally.

**Review by SeanTAllen — COMMENTED** [review]
Would it be reasonable to say based on the number of things with 20 minute time outs that if 1 or 2 or these hang that not all of them will be run before the CI run is terminated?

**Comment by SeanTAllen** [comment]
@AntonioND rather than using expect, would it be reasonable to merely check the output value like we do with existing tests?

File: samples/containers/encrypted/test.sh
Comment (AntonioND) [reply samples/containers/encrypted/test.sh:14]: I'm going to create an issue for this, and for the openmp and nodejs broken samples.

File: samples/containers/encrypted/test.sh
Comment (AntonioND) [reply samples/containers/encrypted/test.sh]: This one doesn't take as long, sorry. The problem was the ``basic/attack`` sample, that can easily take 10-15 minutes in my VM. I just copy pasted the ``test.sh`` script from that folder and forgot to change it.

Comment (SeanTAllen) [reply samples/containers/encrypted/test.sh]: I could be wrong but I think you have 20 minutes for most timeouts that I saw in this PR. I would suggest lowering them if it isn't needed for them.

**Comment by AntonioND** [comment]
> @AntonioND rather than using expect, would it be reasonable to merely check the output value like we do with existing tests?

That's not possible in all cases. For example, the ``language/java`` sample doesn't exit after it's done printing the demo message, I have to wait until it's printed and then send a ``"^C"`` for the test to finish. For the ``basic/attack`` sample you have to wait until the initial process is running and it prints "I'm ready to be attacked", attack it from a second process, and send a ``"\n"`` to the initial one to end. I tried doing this by having a countdown in the initial process and launching the second one at the right time, but it wasn't reliable at all. It is pretty reliable with expect.

EDIT: Maybe that behaviour of the java sample is a bug that could be added to the issue I'm going to create to fix the other samples...

**Comment by SeanTAllen** [comment]
>  It is pretty reliable with expect.

That sounds like it is flakey. Are you saying it should be expected to fail sometimes with expect?

File: samples/languages/dotnet/README.md
Comment (AntonioND) [reply samples/languages/dotnet/README.md:15]: Other samples are as vague as this, and I didn't want to spend too much time on changing readmes in this PR.

I think this should be done as part of https://github.com/lsds/sgx-lkl/issues/332, which is a bigger task. It would be nice to get rid of the Makefiles, or to make then easier to understand. But, definitely, all samples need to be cleaned up a bit.

**Comment by AntonioND** [comment]
> That sounds like it is flakey. Are you saying it should be expected to fail sometimes with expect?

I can't be 100% sure it is always going to work, specially after seeing that some samples (like the java one) don't exit when I would expect them to.

Regarding the ones I'm testing with expect, I've never seen them fail so far. Of course, if someone changes the text output of the demo code the expect script will break, that's what I was thinking about when I wrote "pretty reliable".

In general I agree that expect is not a nice thing to use, but in the case of ``basic/attack`` I don't know how to do it in a different way, and in the case of ``languages/java``, assuming that the hangup is not a bug (I don't know the expected behaviour after printing the test output), I don't know how to make the test end either.

File: samples/basic/attack/read_memory.sh
Comment (AntonioND) [reply samples/basic/attack/read_memory.sh]: Done.

**Comment by AntonioND** [comment]
> Would it be reasonable to say based on the number of things with 20 minute time outs that if 1 or 2 or these hang that not all of them will be run before the CI run is terminated?

I've also fixed this, now there is only one sample that has that timeout. It's not needed when I run the CI, but it is needed when I run the test in my VM.

File: samples/containers/encrypted/test.sh
Comment (AntonioND) [reply samples/containers/encrypted/test.sh]: Done.

File: samples/containers/redis/README.md
Comment (AntonioND) [reply samples/containers/redis/README.md]: Done.

File: samples/languages/java/test.sh
Comment (AntonioND) [reply samples/languages/java/test.sh]: Fixed.

File: samples/languages/java/README.md
Comment (AntonioND) [reply samples/languages/java/README.md]: Fixed.

File: samples/containers/encrypted/test.sh
Comment (AntonioND) [reply samples/containers/encrypted/test.sh:14]: https://github.com/lsds/sgx-lkl/issues/831

**Comment by AntonioND** [comment]
I've left a note in the ``languages/java`` sample that mentions that the sample isn't supposed to need expect scripts, and to remove them when the issue is fixed. I also mention in the comment the GitHub issue that needs to be fixed.

File: samples/basic/helloworld/enclave_config.json
Comment (wintersteiger) [inline samples/basic/helloworld/enclave_config.json]: I think we should keep `enclave_config.json`; it's the "new" way of passing a configuration to the launcher and the samples should show how they're used.

**Review by wintersteiger — APPROVED** [review]
I agree, it would be nice if we could get of `expect`, but I don't have a good alternative solution either.

File: samples/basic/helloworld/enclave_config.json
Comment (AntonioND) [reply samples/basic/helloworld/enclave_config.json]: Makes sense. Restored and modified the Makefile to use it again.

**Comment by AntonioND** [comment]
So @prp mentioned to me that the java sample is failing because the number of ethreads was 1. I've increased the number of ethreads in that sample and now it's working fine and it doesn't need expect scripts. However, the ``basic/attack`` one still needs them (but it's the only sample that does).

**Comment by SeanTAllen** [comment]
LGTM other than the CI failures which I believe are unrelated.

**Review by vtikoo — APPROVED** [review]
LGTM

File: .azure-pipelines/scripts/run_sample.sh
Comment (hukoyu) [inline .azure-pipelines/scripts/run_sample.sh]: Will run samples for run-hw
will be better. It will be better if you change messages with "tests" to "samples" in run_sample.sh

File: .azure-pipelines/scripts/run_sample.sh
Comment (hukoyu) [inline .azure-pipelines/scripts/run_sample.sh]: Same here. Use sample instead of test

File: samples/basic/helloworld/test.sh
Comment (hukoyu) [inline samples/basic/helloworld/test.sh:16]: This is 1 second and it is too short to be default value. May be 60 seconds is better.

File: samples/common.mk
Comment (hukoyu) [inline samples/common.mk]: Do you need to redefine common.mk? Can you just include ../tests/common.mk ?

File: samples/common.sh
Comment (hukoyu) [inline samples/common.sh:4]: These are already exported in Makefile. Do we need to export again? Also SGXLKL_SETUP_TOOL exists in common.sh and not in common.mk just FYI

File: samples/containers/encrypted/test.sh
Comment (hukoyu) [inline samples/containers/encrypted/test.sh:18]: 1 second is too small

File: samples/containers/redis/run-redis-client.sh
Comment (hukoyu) [inline samples/containers/redis/run-redis-client.sh]: Sample succeeded

File: samples/containers/redis/test.sh
Comment (hukoyu) [inline samples/containers/redis/test.sh:24]: 1 second too short

File: samples/languages/dotnet/test.sh
Comment (hukoyu) [inline samples/languages/dotnet/test.sh:16]: 1 second too short

File: samples/languages/java/test.sh
Comment (hukoyu) [inline samples/languages/java/test.sh:16]: 1 second is too short

File: samples/languages/python/test.sh
Comment (hukoyu) [inline samples/languages/python/test.sh:31]: 1 second is too short

**Review by hukoyu — APPROVED** [review]
@AntonioND overall looks good. I provided some feedback. After addressing them this PR is good to be merged.

File: samples/basic/helloworld/test.sh
Comment (AntonioND) [reply samples/basic/helloworld/test.sh:16]: No, this means "default". This is how it works in the case of the Makefiles of the tests. If there is no target called ``gettimeout``, ``make`` will return 1. If the target is there, it prints the timeout length.

Take a look at ``tests/ltp/batch.mk``, target ``gettimeout``, for example.

File: samples/containers/redis/test.sh
Comment (AntonioND) [reply samples/containers/redis/test.sh:24]: Same comment as before.

File: samples/languages/java/test.sh
Comment (AntonioND) [reply samples/languages/java/test.sh:16]: Same comment as before.

File: samples/languages/python/test.sh
Comment (AntonioND) [reply samples/languages/python/test.sh:31]: Same comment as before.

File: samples/languages/dotnet/test.sh
Comment (AntonioND) [reply samples/languages/dotnet/test.sh:16]: Same comment as before.

File: samples/containers/encrypted/test.sh
Comment (AntonioND) [reply samples/containers/encrypted/test.sh:18]: Same comment as before.

File: samples/common.sh
Comment (AntonioND) [reply samples/common.sh:4]: Some samples need this in the bash script because not all of the commands are run in the Makefile. For example, ``SGXLKL_SETUP_TOOL`` is needed in ``containers/redis/test.sh``. A couple of them are needed in ``languages/python/test.sh``.

I only added ``SGXLKL_SETUP_TOOL`` to the bash script because it's not needed in the Makefiles.

File: .azure-pipelines/scripts/run_sample.sh
Comment (AntonioND) [reply .azure-pipelines/scripts/run_sample.sh]: Done.

File: .azure-pipelines/scripts/run_sample.sh
Comment (AntonioND) [reply .azure-pipelines/scripts/run_sample.sh]: Done.

File: samples/common.mk
Comment (AntonioND) [reply samples/common.mk]: Done.

File: samples/containers/redis/run-redis-client.sh
Comment (AntonioND) [reply samples/containers/redis/run-redis-client.sh]: Done.

**Comment by AntonioND** [comment]
I'm going to merge this PR now after addressing the last few comments. I'm still going to be working on the samples for a while, so I'm happy to push a PR with more changes if needed.

**Pull request merged**
